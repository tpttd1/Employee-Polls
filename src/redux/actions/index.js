import { ADD_QUESTION, STORAGE_QUESTIONS, LOGIN_USER_SUCCESS, STORAGE_USERS } from "../../constants";
import { _getQuestions, _getUsers } from "../../data/_DATA";

const saveQuestions = (questions) => ({
  type: STORAGE_QUESTIONS,
  questions,
});

export const storeQuestions = (questions) => (dispatch) => {
  dispatch(saveQuestions(questions));
};

export const getCurrentQuestions = () => (dispatch) => {
  _getQuestions().then((data) => {
    const questions = Object.values(data);
    questions.sort((a, b) => {
      return b.timestamp - a.timestamp;
    });
    dispatch(storeQuestions(questions));
  });
};

const addQuestion = (question) => ({
  type: ADD_QUESTION,
  question,
});

export const createNewPoll = (question) => (dispatch) => {
  dispatch(addQuestion(question));
};

export const getCurrentUsers = () => (dispatch) => {
  _getUsers().then((data) => {
    dispatch(storeUsers(Object.values(data)));
  });
};

const saveUsers = (users) => ({
  type: STORAGE_USERS,
  users,
});

export const storeUsers = (users) => (dispatch) => {
  dispatch(saveUsers(users));
};

const addUser = (user) => ({
  type: LOGIN_USER_SUCCESS,
  user,
});

export const loginUser = (user) => (dispatch) => {
  dispatch(addUser(user));
};
