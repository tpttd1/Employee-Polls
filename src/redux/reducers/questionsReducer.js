import { ADD_QUESTION, STORAGE_QUESTIONS } from "../../constants";

const INIT_STATE = {
  questions: [],
  loading: false,
};

export default function questionsReducer(state = INIT_STATE, action) {
  switch (action.type) {
    case ADD_QUESTION:
      return {
        ...state,
        loading: false,
        questions: [...state.questions, action.questions],
      };
    case STORAGE_QUESTIONS:
      return { ...state, loading: false, questions: action.questions };
    default:
      return { ...state };
  }
}
