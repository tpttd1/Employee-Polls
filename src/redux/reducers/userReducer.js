import {
  LOGIN_USER_SUCCESS,
  LOGOUT_USER,
  STORAGE_USERS,
} from "../../constants";

const INIT_STATE = {
  user: null,
  users: [],
  loading: false,
};

export default function userReducer(state = INIT_STATE, action) {
  switch (action.type) {
    case LOGIN_USER_SUCCESS:
      return { ...state, user: action.user };
    case STORAGE_USERS:
      return { ...state, users: action.users };
    case LOGOUT_USER:
      return { ...state, user: null };
    default:
      return { ...state };
  }
}
