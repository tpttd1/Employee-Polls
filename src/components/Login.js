import { connect } from "react-redux";
import { useNavigate } from "react-router-dom";
import { loginUser } from "../redux/actions";

function Login(props) {
  const navigate = useNavigate();

  const onSubmit = (e) => {
    e.preventDefault();
    const formData = new FormData(e.target);
    const username = formData.get("username");
    const password = formData.get("password");
    const user = { username, password };
    props.dispatch(loginUser(user));
    navigate("/");
  };

  return (
    <div className="login-page">
      <form className="form-login" onSubmit={onSubmit}>
        <div className="img-container">
          <img src="login-icon.png" alt="Avatar" className="avatar" />
        </div>
        <div className="form-input">
          <label htmlFor="username">Username</label>
          <input
            type="text"
            placeholder="Enter username"
            name="username"
            minLength={3}
            className="input input-username"
            required
          />
          <label htmlFor="password">Password</label>
          <input
            type="password"
            placeholder="Enter password"
            name="password"
            minLength={6}
            className="input input-password"
            required
          />
          <button id="btn-login" type="submit">
            Login
          </button>
        </div>
      </form>
    </div>
  );
}

export default connect((state) => state)(Login);
