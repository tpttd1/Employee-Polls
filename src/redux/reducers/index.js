import { combineReducers } from "redux";

import auth from "./userReducer";
import question from "./questionsReducer";

export default combineReducers({
  auth,
  question,
});
