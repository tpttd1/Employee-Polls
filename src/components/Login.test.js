/* eslint-disable testing-library/render-result-naming-convention */
import { fireEvent, render, screen } from "@testing-library/react";
import Login from "./Login";
import { applyMiddleware, legacy_createStore } from "redux";
import reducers from "../redux/reducers";
import thunk from "redux-thunk";
import { Provider } from "react-redux";

const store = legacy_createStore(reducers, applyMiddleware(thunk));

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => (jest.fn())
}));

describe("Test component Login", () => {
  it("Test fire event", () => {
    const component = render(
      <Provider store={store}>
        <Login />
      </Provider>
    );

    fireEvent.change(screen.getByTestId("test-select"), {
      target: { value: "sarahedo" },
    });

    expect(store.getState().auth.user).toEqual(undefined);
    expect(component).toMatchSnapshot();
  });
});
