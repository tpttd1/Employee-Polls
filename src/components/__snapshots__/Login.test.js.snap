// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test component Login Test fire event 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="login-page"
      >
        <div
          id="wrapper-login-as"
        >
          <p
            id="title-login-as"
          >
            Login as
          </p>
          <select
            data-testid="test-select"
            id="select-user"
          >
            <option
              class="item-user"
              value="none"
            >
              Select an user
            </option>
          </select>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="login-page"
    >
      <div
        id="wrapper-login-as"
      >
        <p
          id="title-login-as"
        >
          Login as
        </p>
        <select
          data-testid="test-select"
          id="select-user"
        >
          <option
            class="item-user"
            value="none"
          >
            Select an user
          </option>
        </select>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
